data-service/src/main/java/uk/gov/gchq/palisade/service/data/web/DataController.java

```diff
@@ -73,36 +68,31 @@ public DataController(final DataService dataService,
-        DataReaderRequest readerRequest = dataService.authoriseRequest(dataRequest)
-                .thenApply(maybeReadRequest -> maybeReadRequest
-                        .orElseThrow(() -> new ForbiddenException(
-                                String.format("The token '%s' is not authorised to access the leafResource '%s'", dataRequest.getToken(), dataRequest.getLeafResourceId()))))
-                .join();
 
-        StreamingResponseBody stream = (OutputStream outputStream) -> {
-            Pair<AtomicLong, AtomicLong> recordsAudit = dataService.read(readerRequest, outputStream);
-            AuditSuccessMessage successMessage = auditService.createSuccessMessage(dataRequest, readerRequest, recordsAudit.getFirst().get(), recordsAudit.getSecond().get());
-            auditService.auditSuccess(dataRequest.getToken(), successMessage);
-        };
+        HttpStatus httpStatus = HttpStatus.ACCEPTED;
+        StreamingResponseBody stream = null;
 
-        LOGGER.info("Streaming response: {}", stream);
-        return new ResponseEntity<>(stream, HttpStatus.OK);
-    }
+        //first with the client information about the request, retrieve the authorised resource information
+        AuditableAuthorisedDataRequest auditableAuthorisedDataRequest = auditableDataService.authoriseRequest(dataRequest).join();
+        AuditErrorMessage authorisationErrorMessage = auditableAuthorisedDataRequest.getAuditErrorMessage();
 
-    /**
-     * This method will add a {@link uk.gov.gchq.palisade.data.serialise.Serialiser} to the
-     * {@link DataService} using the details provided in a yaml file.
-     */
-    @EventListener(ApplicationReadyEvent.class)
-    public void initPostConstruct() {
-        // Add serialiser to the data-service
-        LOGGER.info("Prepopulating using serialiser config: {}", serialiserConfig.getClass());
-        serialiserConfig.getSerialisers().stream()
-                .map(StdSerialiserPrepopulationFactory::build)
-                .peek(entry -> LOGGER.debug(entry.toString()))
-                .forEach(entry -> dataService.addSerialiser(entry.getKey(), entry.getValue()));
-    }
+        if (authorisationErrorMessage != null) {
+            LOGGER.error("Error occurred processing the authoriseRequest for  {}", authorisationErrorMessage);
+            httpStatus = (HttpStatus.INTERNAL_SERVER_ERROR);
+            auditMessageService.auditMessage(new TokenMessagePair(dataRequest.getToken(), authorisationErrorMessage)).join();
+        } else {
+            // Create a consumer of the REST response's OutputStream, writing resource data to it
+            stream = (OutputStream outputStream) -> {
+                AuditableDataResponse auditableDataResponse = auditableDataService.read(auditableAuthorisedDataRequest, outputStream).join();
+                auditMessageService.auditMessage(new TokenMessagePair(dataRequest.getToken(), auditableDataResponse.getAuditSuccessMessage())).join();
```

Comment:
Any particular reason that these join()s are here? This is not a reactive web controller, so rationally we would want the audit to complete prior to the stream returning control to the client so why not put them in auditMessage() and allow this method to block.

Reply:
Furthermore, you probably want some `.exceptionally` clause on the `service.read` such that we can guarantee an audit as best we can

Reply:
Done, read method blocks until it is finished.

Reply:
@dev930018 , I think you mean the auditMessage method on the AuditMessageService.  If that is the case, we would be talking about local logging if Kafka stream failed.  Is that right?